plugins/module_utils/data_merge.py

```diff
@@ -0,0 +1,164 @@
+                    the result.  If an element in expected data as different
+                    value or in not present in the base data, it will be set or
+                    added in the result.
+        - absent  : Ensure that an element is absent in the final result. For
+                    the dict, if a key is present in the current dict and in
+                    the expected dict but has different value, the current
+                    value will be kept in the result.
+        """
+        if value not in ['identic', 'present', 'absent']:
+            raise ValueError('`merge_type` can be only one of `identic`, `present` or `absent`')
+        self._merge_type = value
+
+    @property
+    def list_diff_type(self):
+        """
+        Descibe the mean of merging data for a list.
+        """
+        return self._list_diff_type
+
+    @list_diff_type.setter
+    def list_diff_type(self, value):
+        """
+        Descibe the mean of merging data for a list.
+        Can be one of :
+        - value :   Comparing elements in the list by their values. Search if
+                    an element in expect list is present or absent, depending
+                    on the MergType, to the current list.  If it not the case
+                    it will be added of removed. This not check list
+                    recursively, only the first level is checked.
+        - index :   Comparing elements in the list by their index. If
+                    `self.merge_type` is `present`, it will check if the value
+                    of all elements, in the current list as the same value as
+                    the element in the expected list with the same index, and
+                    if it not the case, the value of the element in the
+                    expected list will be set at this index position in the
+                    result list. If `self.merge_type` is `absent`, it will
+                    check if the value of an element in the current list as the
+                    same value as the element with the same index in the
+                    expected list and if it's the case, the element will not be
+                    present in the result.
+        """
+        if value not in ['value', 'index']:
+            raise ValueError('`list_diff_type` can be only one of `value` or `index`')
+        self._list_diff_type = value
+
+    def get_new_merged_data(self, current, expected):
+        # type: (dict | list, dict | list) -> dict | list
+        """
+        Getting the merge of two element if it can't be sure that they have the
+        same type.
+        """
+        if not isinstance(current, (dict, list)) or not isinstance(expected, (dict, list)):
+            raise ValueError('Only dict or list can be merged, {0} and {1} given.'.format(type(current), type(expected)))
+        if self.merge_type == 'identic':
+            return deepcopy(expected)
```

Comment:
The filter use `DataMerge` that I designed to be reusable in various context, not only in this filter.

I'm afraid that if I remove the `deepcopy` in `get_new_merged_data`, `get_new_merged_dict` and `get_new_merged_list` we got unexpected behaviours later.

Firstly, we get a problem when we use `merge_type='absent`. `get_new_merged_dict` call itself recursively for each `dict` element that it contains. If we do not return `deepcopy` for the result we modifing the original object by removing some element on it, and we get problems when we try to continus to iterating over its elements (concretely, it raises `RuntimeError: dictionary changed size during iteration`).

Secondly, when we call `get_new_merged_data`, `get_new_merged_dict` and `get_new_merged_list`, we expect to get a new object. If we remove `deepcopy` it possible to get one of `current` or `expected` object, depending on the used params, modified by the call of the method.  If I need to reuse later one of `current` or `expected` I can't because the original value has been replaced by the result. It's bit tricky...

I don't see very clearly what ansible-core objects can be passed to the filter. Please, can you give me some sample to help me that understand this. Maybe I can treat the problem in `merge_data`, the function called by the filter, avoiding modifying anyone in `DataMerge`.

Reply:
> I'm afraid that if I remove the `deepcopy` in `get_new_merged_data`, `get_new_merged_dict` and `get_new_merged_list` we got unexpected behaviours later.

I do not mean simply removing `deepcopy` without replacement, but writing code which handles this correctly.

> I don't see very clearly what ansible-core objects can be passed to the filter. Please, can you give me some sample to help me that understand this. Maybe I can treat the problem in `merge_data`, the function called by the filter, avoiding modifying anything in `DataMerge`.

For example `hostvars` is of type `ansible.vars.hostvars.HostVars`, and `hostvars[inventory_hostname]` is of type `ansible.vars.hostvars.HostVarsVars`. Also lookups and filters can return arbitrary Python objects.

You should probably look at how `merge_hash` is implemented in lib/ansible/utils/vars.py in https://github.com/ansible/ansible/